src/net/managers.rs

```diff
@@ -0,0 +1,141 @@
+    /// `SocketManager` can decide when to destroy the connection if two many warnings are propagated from the same connection in a short amount of time.
+    Warning(String), // TODO: is it enought? or maybe we need more fields?
+}
+
+/// It abstracts pure UDP packets, and allows to implement Connected/Disconnected states.
+/// This table summary shows where exactly ConnectionManager sits in between different layers.
+/// | Abstraction layer | Capabilities                                                                                                                                                         |
+/// |-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+/// | Application       | Can receive these events: Created->Connected(data)->Packet(data)->Disconnected(reason)->Destroyed(reason). Can send these events: Connect(data)->Packet->Disconnect. |
+/// | ConnectionManager | Receives all, except user packets, and can report state updates, and generate new packets via `update` method                                                        |
+/// | Laminar           | Adds/Removes headers to packets, so that it could provides reliability, ordering, fragmentation, etc.. capabilities.                                                 |
+/// | ConnectionManager | May change raw incoming and outgoing bytes to apply encryption, compression, etc.                                                                                    |
+///
+/// It tries to maintain a valid connection state, and it can't decide when to destroy itself, only when it changes to disconnected, it will be destroyed later.
+/// From the point of view of connection manager, laminar's header + payload is interpreted as user data.
+/// Distinction between user packet and the protocol-specific packet is encoded in laminar's packet header.
+/// Preprocess/Postprocess and Update methods always accept temporary buffer of size `Config.receive_buffer_max_size` that can be used as output.
+pub trait ConnectionManager: Debug + Send {
+    /// When the instance of the connection manager is created, the `update` method will be called, before any other method.
+    /// This function should be called frequently, even if there are no packets to send or receive.
+    /// It will always be called last, after all, other methods are called, in the main laminar`s loop.
+    /// It can generate all kinds of packets: heartbeat, user or connection protocol packets.
+    /// (maybe heartbeat functionality should be moved here?)
+    /// It will be called in the loop as long, as it returns any results. E.g. `connect` method may generate multiple results: change state and send packet.
+    fn update<'a>(
+        &mut self,
+        buffer: &'a mut [u8],
+        time: Instant,
+    ) -> Option<Result<Either<GenericPacket<'a>, ConnectionState>, ConnectionManagerError>>;

```

Comment:
This return value terrifies me. Really smells like this method is trying to do way too many things. There is most definitely a better structure here.

Reply:
Good point, it doesn't look very intuitive. I updates to this:
```rust
/// Events that are generated by `ConnectionManager`s update method.
#[derive(Debug)]
pub enum ConnectionManagerEvent<'a> {
    /// Generated new packet, that will be sent immediatelly.
    NewPacket(GenericPacket<'a>),
    /// `ConnectionState` state changed.
    NewState(ConnectionState),
    /// Error occured within `ConnectionManager`
    Error(ConnectionManagerError)
}
```
and return type is now `Option<ConnectionManagerEvent<'a>>`

Reply:
That already looks much better.